server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,249 @@
+patterns for identifying a workflow execution and carrying arbitrary metadata
+(Header) and payloads (Payloads). The request type for this function includes a
+`result_access_style` property indicating whether the caller intends to poll for
+a result later (via `PollUpdateWorkflowResponse`) or wants the result to be
+delivered "inline" as part of the `UpdateWorkflowResponse` that is returned from
+the `UpdateWorkflow` gRPC call. In the case where the client intends to poll,
+the RPC will return as soon as the UpdateID is generated (in such cases, the
+`UpdateWorkflowResponse` may contain the actual UpdateResponse data or may be
+empty with the `UPDATE_WORKFLOW_RESULT_TYPE` flag set to
+`UPDATE_WORKFLOW_RESULT_TYPE_IN_PROGRESS`. If the client wants an inline response value,
+the RPC will block until the response value (or error) is available or the gRPC
+timeout is hit. In either case, the gRPC timeout for this call is server-side
+capped at low value, indicating the intent that this call either complete
+quickly or at least be validated quickly and the client will switch to polling.
+The `request_id` property uniquely identifies the logical request (_not_ the
+issuance of the request) and is intended for use as an idempotency token.
+
+As `UpdateWorkflow` executes workflow code in all cases it is recommended that
+this function be excluded from rpc latency metrics aggregation.
+
+The second function, `PollUpdateWorkflowResponse` is used when a client has
+already issued an `UpdateWorkflow` request and has the resulting `update_id`.
+That `update_id` can be provided to re-attach to an in-flight or completed
+Update and obtain the associated UpdateResult in the form of an
+`UpdateWorkflowResponse`.
+
+
+```protobuf
+service WorkflowService {
+    rpc UpdateWorkflow(UpdateWorkflowRequest) returns (UpdateWorkflowResponse){}
+    rpc PollUpdateWorkflowResponse(PollUpdateWorkflowResponseRequest) returns (UpdateWorkflowResponse){}
+}
+
+message UpdateWorkflowRequest {
+    temporal.api.common.v1.Header header = 1;
+    string request_id = 2;
+    string namespace = 3;
+    temporal.api.common.v1.WorkflowExecution execution = 4;
+    string update_name = 5;
+    temporal.api.common.v1.Payloads input = 6;
+    UpdateWorkflowResultAccessStyle result_access_style = 7;
+}
+
+message UpdateWorkflowResponse {
+    string update_id = 1;
+    UpdateWorkflowResultType result_type = 2;
+    UpdateWorkflowSuccess success = 3;
+    UpdateWorkflowFailure failure = 4;
+}
+
+enum UpdateWorkflowResultAccessStyle {
+    UPDATE_WORKFLOW_RESULT_ACCESS_STYLE_UNSPECIFIED = 0;
+    UPDATE_WORKFLOW_RESULT_ACCESS_STYLE_INLINE = 1;
+    UPDATE_WORKFLOW_RESULT_ACCESS_STYLE_POLL = 2;
+}
+
+enum UpdateWorkflowResultType {
+    UPDATE_WORKFLOW_RESULT_TYPE_UNSPECIFIED = 0;
+    UPDATE_WORKFLOW_RESULT_TYPE_IN_PROGRESS = 1;
+    UPDATE_WORKFLOW_RESULT_TYPE_SUCCESS = 2;
+    UPDATE_WORKFLOW_RESULT_TYPE_FAILURE = 3;
+}
+
+message PollUpdateResponseRequest {
+    string update_id = 1; 
```

Comment:
I am not quite sure how server is able to route the poll request to a specific history host, if needed, or find the workflow containing the update response just from the `update_id`. 

If the `update_id` contains extra information like target ns, workflowID, runID, would it be better to return an opaque binary token instead of a string?

Reply:
> ... would it be better to return an opaque binary token instead of a string?

I think you're right. I'll switch to `google.protobuf.BytesValue` or a `temporal.api.common.v1.Payload`

Reply:
Or simply `bytes`? Looks like most tokens we returned have `bytes` type, so might be good to be consistent? e.g. https://github.com/temporalio/api/blob/master/temporal/api/workflowservice/v1/request_response.proto#L258

I don't have strong opinion though.

Reply:
I'd prefer the wrapper types but since we're already using raw `bytes` elsewhere I don't think the benefit is enough to justify being inconsistent.

Reply:
Marking unresolved just to get my comment seen (feel free to mark resolved after).

While I understand this is a server provided identifier, I wonder if there are other concerns with packing a custom binary encoded value into the identifier. But even if we do, arguably a UI-visible identifier has value anyways. Does one expect a large blob to be base64'd by the frontend here? Or in logs when one is handle the update and wants to uniquely identify it?

I don't mind leaving as is (sounds like we may have to if we're not willing to do state lookup on our side), just want to make sure we consider the other benefits of short, readable, arbitrary identifiers over encoded blob identifiers. (feel free to resolve conversation)